File: src/io/github/deltarays/discordconsole/LogAppender.java
package io.github.deltarays.discordconsole;

import org.apache.commons.lang.StringUtils;
import org.apache.logging.log4j.core.LogEvent;
import org.apache.logging.log4j.core.appender.AbstractAppender;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.text.SimpleDateFormat;
import java.util.*;

public class LogAppender extends AbstractAppender {
    DiscordConsole main;
    Boolean startupDone = false;
    Queue<String> msgs = new LinkedList<>();

    public LogAppender(DiscordConsole main) {
        super("DiscordConsoleLogAppender", null, null);
        start();
        this.main = main;
        Timer timer = new Timer();
        timer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                try {
                    StringBuilder message = new StringBuilder();
                    for (String ln = msgs.poll(); ln != null; ln = msgs.poll()) {
                        if (message.length() + ln.length() > 1999) {
                            main.sendDiscordMessage(message.toString());
                            message = new StringBuilder();
                        }
                        message.append(ln).append("\n");
                    }

                    if (StringUtils.isNotBlank(message.toString().replace("\n", ""))) {
                        main.sendDiscordMessage(message.toString());
                    }
                    main.updateBotStatus();
                } catch (Exception e) {
                    main.getLogger().severe("Error in sending console logs to channel!");
                    e.printStackTrace();
                }
            }
        }, 0, (Math.max(main.getConfig().getInt("channelRefreshRate"), 1)) * 1100);
    }

    @Override
    public void append(LogEvent ev) {
        Method method = null;
        try {
            method = LogEvent.class.getMethod("toImmutable");

        } catch (NoSuchMethodException ignored) {
        }
        if (method != null) {
            try {
                ev = (LogEvent) method.invoke(ev);
            } catch (IllegalAccessException | InvocationTargetException ignored) {
            }
        }
        Long timeMillis = null;
        method = null;
        try {
            method = LogEvent.class.getMethod("getTimeMillis");
        } catch (NoSuchMethodException e) {
            try {
                timeMillis = (Long) LogEvent.class.getMethod("getMillis").invoke(ev);
            } catch (Exception ignored) {
            }
        }
        if (method != null) {
            try {
                timeMillis = (Long) method.invoke(ev);
            } catch (IllegalAccessException | InvocationTargetException ignored) {
            }
        }
        if (timeMillis == null) timeMillis = new Date().getTime();
        LogEvent event = ev;
        String message = event.getMessage().getFormattedMessage();
        boolean sendStartupMessages = !main.getConfig().isSet("sendStartupMessages") || main.getConfig().getBoolean("sendStartupMessages");
        if (message.toLowerCase().contains("reload") && message.toLowerCase().contains("complete")) startupDone = true;
        if (startupDone || sendStartupMessages) {
            SimpleDateFormat formatter = new SimpleDateFormat("HH:mm:ss");
            message = String.format("[%s] [%s/%s] %s", formatter.format(new Date(timeMillis)), event.getThreadName(), event.getLevel().toString(), message);
            message = message.replaceAll("\\[m|\\[([0-9]{1,2}[;m]?){3}|\u001b+", "").replaceAll("\\x1b\\[[0-9;]*[A-Za-z]]*", "").replace("_", "\\_").replace("*", "\\*").replace("~", "\\~").replaceAll("([&§])[0-9a-fklmnor]", "");
            msgs.add(message);
        }
    }
}

File: src/main/kotlin/me/deltarays/discordconsole/logging/LogAppender.kt
package me.deltarays.discordconsole.logging

import me.deltarays.discordconsole.DiscordConsole
import me.deltarays.discordconsole.Utils
import me.deltarays.discordconsole.discord.DiscordChannel
import org.apache.logging.log4j.core.LogEvent
import org.apache.logging.log4j.core.appender.AbstractAppender
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.regex.Pattern

class LogAppender(private var plugin: DiscordConsole) :
    AbstractAppender("DiscordConsoleAppender", null, null, false) {
    init {
        isAttached = true
        this.start()
    }

    override fun append(event: LogEvent?) {
        val logEvt: LogEvent?
        val method = try {
            LogEvent::class.java.getMethod("toImmutable")
        } catch (e: NoSuchMethodException) {
            null
        }
        logEvt = if (method != null) {
            try {
                method.invoke(event) as LogEvent?
            } catch (exc: Exception) {
                event
            }
        } else {
            event
        }
        val timeMillis: Long = try {
            LogEvent::class.java.getMethod("getTimeMillis").invoke(logEvt)
        } catch (e: Exception) {
            try {
                LogEvent::class.java.getMethod("getMillis").invoke(logEvt)
            } catch (exc :Exception){
                Date().time
            }
        } as Long
        val message = logEvt?.message?.formattedMessage as String
        for (channel in DiscordChannel.channels) {
            if (channel.types.contains(LogType.CONSOLE)) {
                val channelSection = plugin.getConfigManager().getChannel(channel.id)
                val consoleSection =
                    channelSection.getConfigurationSection("console") ?: channelSection.createSection("console")
                if (consoleSection.getBoolean("send-startup", false) || DiscordConsole.serverHasStartedUp) {
                    val fmt: String = channel.getMessageFormat(LogType.CONSOLE)
                    val formatted =
                        parse(fmt, logEvt.level.name(), message, logEvt.threadName, timeMillis)
                    val filterStr = consoleSection.getString("filter") ?: ""
                    if (filterStr.isNotEmpty()) {
                        val filter = Pattern.compile(filterStr)
                        if (!filter.matcher(formatted).find()) continue
                    }
                    channel.enqueueMessage(formatted)
                }
            }


        }
    }

    companion object {
        var isAttached = false
    }

    private fun parse(str: String, levelName: String, message: String, threadName: String, time: Long): String {
        return Utils.convertPlaceholders(str).replace(Regex("\\{message}", RegexOption.IGNORE_CASE), message)
            .replace(Regex("\\{level}", RegexOption.IGNORE_CASE), levelName)
            .replace(Regex("\\{thread}", RegexOption.IGNORE_CASE), threadName)
            .replace(Regex("\\{date\\[(.*?)]}", RegexOption.IGNORE_CASE)) { e ->
                val dateFormat = DateTimeFormatter.ofPattern(e.groupValues.getOrElse(1) { "hh:mm:ss" })
                dateFormat.format(Instant.ofEpochMilli(time).atOffset(ZoneOffset.UTC))
            }
            // Replaces special characters
            .replace("\\[m|\\[([0-9]{1,2}[;m]?){3}|\u001b+".toRegex(), "")
            .replace("\\x1b\\[[0-9;]*[A-Za-z]]*".toRegex(), "")
            .replace("([&§])[0-9a-fklmnor]".toRegex(), "")
            .replace("_", "\\_").replace("*", "\\*")
            .replace("~", "\\~")
            .replace("|", "\\|")
            .replace(">", "\\>")
            .replace("`", "\\`")
    }
}
